### PR TITLE
e2e: add delete apiserver test

### PR DESCRIPTION
**Pull request by pbx0**
Also adds a work-around for hashicorp/terraform#1885

### CODE REVIEW

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
Tests are passing. The most recent pluton run had a flake, but the tests that actually run the code in this PR passed twice in a row now.

File: e2e/etcdscale_test.go [hunk 1]
```diff
@@ -132,8 +132,8 @@ func resizeSelfHostedEtcd(c kubernetes.Interface, size int) error {
 		return nil
 	}
 
-	if err := retry(15, 10*time.Second, podsReady); err != nil {
-		return fmt.Errorf("Waited 150 seconds for etcd to scale: %v", err)
+	if err := retry(31, 10*time.Second, podsReady); err != nil {
+		return fmt.Errorf("Waited 300 seconds for etcd to scale: %v", err)
```
Comment (derekparker) [inline e2e/etcdscale_test.go]: Would be `310`, no?

File: hack/terraform-quickstart/iam.tf [hunk 2]
```diff
@@ -1,6 +1,10 @@
 resource "aws_iam_instance_profile" "bk_profile" {
   name_prefix = "bootkube_e2e_profile"
-  role        = "${aws_iam_role.bk_role.name}"
+  role        = "${aws_iam_role.bk_role.id}"
+
+  provisioner "local-exec" {
+    command = "sleep 90"
```
Comment (derekparker) [inline hack/terraform-quickstart/iam.tf]: Why do we need this sleep?

File: e2e/deleteapi_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+			}
+		}
+	}
+
+	// wait for api-server to go down by waiting until listing pods returns
+	// errors. This is potentially error prone, but without waiting for the
+	// apiserver to go down the next step will return sucess before the
+	// apiserver is ever destroyed.
+	waitDestroy := func() error {
+		_, err := client.CoreV1().Pods(namespace).List(metav1.ListOptions{LabelSelector: "k8s-app=kube-apiserver"})
```
Comment (derekparker) [inline e2e/deleteapi_test.go]: nit: Could just do a simpler request to the version endpoint or something, if you're just checking that the API server is up and reachable.

File: e2e/deleteapi_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		return nil
+	}
+
+	if err := retry(100, 500*time.Millisecond, waitDestroy); err != nil {
+		t.Fatal(err)
+	}
+
+	// wait until api server is back up
+	waitAPI := func() error {
+		_, err := client.CoreV1().Pods(namespace).List(metav1.ListOptions{LabelSelector: "k8s-app=kube-apiserver"})
```
Comment (derekparker) [inline e2e/deleteapi_test.go]: Same nit.

File: e2e/etcdscale_test.go (hunk 1 above)
Comment (pbx0) [inline e2e/etcdscale_test.go]: You would think, but the retry function actually will do N retries and N-1 pauses. So `try -> pause -> try -> pause -> try` It doesn't pause after the last try.

File: hack/terraform-quickstart/iam.tf (hunk 2 above)
Comment (pbx0) [inline hack/terraform-quickstart/iam.tf]: https://github.com/hashicorp/terraform/issues/1885

File: e2e/etcdscale_test.go (hunk 1 above)
Comment (derekparker) [inline e2e/etcdscale_test.go]: Ah, got it. Thanks.

File: hack/terraform-quickstart/iam.tf (hunk 2 above)
Comment (derekparker) [inline hack/terraform-quickstart/iam.tf]: :+1: -- would you mind throwing in a link to that issue?

Comment (pbx0) [inline hack/terraform-quickstart/iam.tf]: I included it in the commit, adding it to PR text as well

File: e2e/deleteapi_test.go (hunk 3 above)
Comment (pbx0) [inline e2e/deleteapi_test.go]: I'm fine with this. Offhand do you know the client-go function to call for this (not obvious to me with a bit of searching other then using the rest-client for which the error handling looks weirder)

Comment (pbx0) [inline e2e/deleteapi_test.go]: done

File: e2e/deleteapi_test.go (hunk 4 above)
Comment (pbx0) [inline e2e/deleteapi_test.go]: done

**Comment by pbx0** [comment]
rktbot run e2e

**Comment by pbx0** [comment]
@derekparker PTAL should be ready to go.

**Pull request merged**